### ISSUE TITLE
sign on some block = clean sign

### ISSUE THREAD

**Issue by marius851000** [cuberite/cuberite#3464]
Client version:   1.10.2
Server OS:        Windows 8
build id:        414

When we place a sign on snow and write on it, the sign stays clean.

[i'm french]
[slightly edited by madmaxoft]

**Comment by madmaxoft** [comment]
What's a pancart?

**Comment by QuImUfu** [comment]
googe translator says:
pancart = sign

**Comment by marius851000** [comment]
I'm not very good in english, sorry, I modify the original message. thanks

**Comment by madmaxoft** [comment]
It's alright, as long as we can understand each other :)
Now this smells like a serious bug with snow blocks. I'd like someone to verify similar behavior with chests - place a chest onto snow, try to put some items in it, close it and check that they are there when you open it, and finally break the chest block, see if all the contents are spilt as pickups. Other block entities may misbehave as well. Any testing volunteers?

**Comment by QuImUfu** [comment]
I can confirm some problems with signs. 
A slightly more accurate description of the problem mentioned above:
When you place a sign replacing something (eg. a layer of snow (or tall grass, if it works at all)) and write something on it, it does not save the text written on it. Nearly (double plants do not work at all) all other blocks work fine.

Another problem i encountered, maybe related:
When placing signs replacing something else (eg. tall grass, farn) they very often get destroyed immediately and drop two signs, making it possible to duplicate signs. On the rare occasion they replace the block as intended the first problem occurs.

I am not able to reproduce these problems with any other block i have tested so far.

**Comment by QuImUfu** [comment]
ohh i think i got the problem:
it sets the text for the sing at the position the sign would be placed if the block would not be replaceable.
Why do i think this happens:
when you place a sign above one layer of snow, then place another sign replacing this layer of snow, the edit gui does edit the previously placed sign, not the sign you just have placed.

**Comment by QuImUfu** [comment]
Ok, code wise i think it is caused by ItemSign.h, line 38. 
The AddFaceDirection() function gets called even if the block clicked on did ignore build collision and was therefor replaced.

**Comment by QuImUfu** [comment]
Replacing this in ItemSign.h:
````cpp
override
	{
		// If the regular placement doesn't work, do no further processing:
		if (!super::OnPlayerPlace(a_World, a_Player, a_EquippedItem, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ))
		{
			return false;
		}

		// After successfully placing the sign, open the sign editor for the player:
		AddFaceDirection(a_BlockX, a_BlockY, a_BlockZ, a_BlockFace);
		a_Player.GetClientHandle()->SendEditSign(a_BlockX, a_BlockY, a_BlockZ);
		return true;
	}
````
with this:
````cpp
override
	{
		// Check if placing on something ignoring build collision to edit the correct sign later on:
		BLOCKTYPE ClickedBlock;
		NIBBLETYPE ClickedBlockMeta;
		a_World.GetBlockTypeMeta(a_BlockX, a_BlockY, a_BlockZ, ClickedBlock, ClickedBlockMeta);
		bool replaced = BlockHandler(ClickedBlock)->DoesIgnoreBuildCollision() || BlockHandler(ClickedBlock)->DoesIgnoreBuildCollision(&a_Player, ClickedBlockMeta);
		
		// If the regular placement doesn't work, do no further processing:
		if (!super::OnPlayerPlace(a_World, a_Player, a_EquippedItem, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ))
		{
			return false;
		}
		
		
		// After successfully placing the sign, open the sign editor for the player,
		// use replaced bool to make sure we edit the right sign:
		if (replaced)
		{
			a_Player.GetClientHandle()->SendEditSign(a_BlockX, a_BlockY, a_BlockZ);
		}
		else
		{
			AddFaceDirection(a_BlockX, a_BlockY, a_BlockZ, a_BlockFace);
			a_Player.GetClientHandle()->SendEditSign(a_BlockX, a_BlockY, a_BlockZ);	
		}
		return true;
	}
````
does fix the issue.

**Comment by madmaxoft** [comment]
Great analysis and good fix. Could you make it into a PR?

I'd prefer the second part of the fixed code to have the `SendEditSign` call outside the `if` block, because it is the same code for both branches. Also, the name `replaced` is somewhat vague, we prefer directly "boolean" names, such as `isReplacing` or even `isReplacingClickedBlock` (note that the "is" part makes it clear that it's a bool).

**Comment by Seadragon91** [comment]
Fixed in #3469.
